text/0000-glimmer-components.md

```diff
@@ -0,0 +1,835 @@
+   * `didInsertElement`,
+   * `didReceiveAttrs`
+   * `willRender`
+   * `didRender`.
+
+3. **Rerenders and Updates:**
+   * `didReceiveAttrs`
+   * `didUpdateAttrs`
+   * `willUpdate`
+   * `didUpdate`
+   * `willRender`
+   * `didRender`
+
+4. **Destruction:**
+   * `willDestroyElement`
+   * `didDestroyElement`
+   * `willDestroy`
+   * `destroy`
+
+
+Many of these hooks have overlapping or redundant functionality, and it's fairly
+confusing when to use which and what the differences are. We can simplify this
+cycle in a number of ways:
+
+* Hooks that run during multiple phases such as `didRender` and
+  `didRecieveAttrs` can be convenient at times, but also add mental overhead and
+  redundancy. We can remove these in favor of clearly delineated hooks which
+  only run during _one_ phase.
+
+* "Bookend" methods (`did*` and `will*`) can be confusing, since they require
+  some specific knowledge of what the "bookended" functionality is. For
+  instance, users almost _always_ want to use `didInsertElement` and
+  `willDestroyElement`, but the existence of their opposite bookends can make
+  this confusing. Additionally, the fact that `didReceiveAttrs` and
+  `didUpdateAttrs` do _not_ have opposing bookends is inconsistent with this
+  pattern.
+
+* Hooks that run immediately _before_ a render such as `didUpdateAttrs` are not
+  particularly useful, because the only functionality they have which
+  post-render methods do not is to modify a bound value without triggering a
+  rerender:
+
+  ```js
+  willUpdateAttrs() {
+    if (this.someBoolArg) {
+      this.set('someBoundValue', true);
+    }
+  }
+  ```
+
+  These types of values can almost always be better expressed using computed
+  properties, which will update when accessed during render, preventing
+  rerenders. Conversely, much imperative logic _must_ be done after a render
+  (manipulating an element directly, for instance), so post render hooks are
+  the most sensible.
+
+* The element based methods can be removed in favor of element modifiers,
+  which are discussed later on.
+
+Based on these considerations, we can reduce these hooks to one per major phase
+of the component's lifecycle:
+
+The **`constructor`** can be used for initial setup of the component. This
+effectively replaces `init`, and allows users to setup state before _any_
+renders occur.
+
+**`didCreate`** runs after the first full render of the component, once the
+component has been fully created. It is closest to `didInsertElement` in legacy
+components, but has been renamed since components no longer have an `element`
+property.
+
+**`didUpdate`** runs after the component has been updated, including all changes
+to `args` and any other changes which trigger a rerender of the component (i.e.
+computed properties being invalidated). It is effectively the same as
+`didUpdate` in legacy components, and can be used for update code.
+
+**`destroy`** is when the component is being destroyed, and can be used for
```

Comment:
Is this invoked depth first, or breadth first, top -> down, or bottom -> up?

Reply:
I see, bellow you do elaborate on the ordering. Should section then link to that, or maybe hint that it will be addressed?